File: util/excel/src/main/java/cn/bestwu/simpleframework/util/excel/ColumnWidths.java
package cn.bestwu.simpleframework.util.excel;

import cn.bestwu.lang.util.CharUtil;
import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

/**
 * @author Peter Wu
 */
public class ColumnWidths {

  private final int maxWidth;
  private final Map<Integer, Double> colWidths = new HashMap<>();

  public ColumnWidths() {
    this(50);
  }

  public ColumnWidths(int maxWidth) {
    this.maxWidth = maxWidth;
  }

  public void put(Integer column, Object val) {
    if (val != null) {
      double width = 0;
      for (char c1 : val.toString().toCharArray()) {
        if (CharUtil.isChinese(c1)) {
          width += 1.5;
        } else {
          width += 1;
        }
      }
      width += +20.0 / 7;
      width = Math
          .max(colWidths.getOrDefault(column, 0.0), width);
      colWidths.put(column, width);
    }
  }

  public Double width(Integer column) {
    return width(column, maxWidth);
  }

  public Double width(Integer column, Integer max) {
    return BigDecimal.valueOf(Math.min(max, colWidths.get(column)))
        .setScale(2, BigDecimal.ROUND_HALF_UP)
        .doubleValue();
  }

}

File: tools/excel/src/main/kotlin/top/bettercode/summer/tools/excel/ColumnWidths.kt
package top.bettercode.summer.tools.excel

import top.bettercode.summer.tools.lang.util.CharUtil.isChinese
import java.math.BigDecimal
import java.math.RoundingMode
import kotlin.collections.set
import kotlin.math.max

/**
 * @author Peter Wu
 */
class ColumnWidths @JvmOverloads constructor(private val maxWidth: Int = 50) {
    private val colWidths: MutableMap<Int, Double> = HashMap()
    fun put(column: Int, `val`: Any?) {
        if (`val` != null) {
            var width = getWidth(`val`)
            width = max(colWidths.getOrDefault(column, 0.0), width)
            colWidths[column] = width
        }
    }

    @JvmOverloads
    fun width(column: Int, max: Int = maxWidth): Double {
        val w = colWidths[column]
        return BigDecimal.valueOf(max.toDouble().coerceAtMost(w ?: 0.0))
                .setScale(2, RoundingMode.UP)
                .toDouble()
    }

    fun clear() {
        colWidths.clear()
    }

    companion object {
        @JvmStatic
        fun getWidth(`val`: Any): Double {
            var width = 0.0
            for (c1 in `val`.toString().toCharArray()) {
                width += if (isChinese(c1)) {
                    1.5
                } else {
                    1.0
                }
            }
            width += 20.0 / 7
            return width
        }
    }
}
